micro_sam/training/peft_sam.py

```diff
@@ -0,0 +1,155 @@
+import math
+from typing import Tuple, List, Union
+
+import torch
+import torch.nn as nn
+
+from segment_anything.modeling import Sam
+
+
+class PEFTBase(nn.Module):
+    """PEFTBase is an interface to implement specific PEFT-based methods.
+    """
+    def __call__(
+        self,
+        rank: int,
+        block: nn.Module
+    ) -> Tuple[nn.Module, List[nn.Module]]:
+        """Returns the attention block by updating the qkv pair per block and the respective linear layers.
+
+        Args:
+            rank: The rank of the per-layer covariance matrix for low-rank adaptation.
+            block: The individual attention block.
+
+        Returns:
+            The attention block.
+            The pair of values for the linear layer.
+        """
+        raise NotImplementedError(
+            "PEFTBase is just a class template. Use a child class that implements the specific PEFT method instead."
+        )
+
+
+class LoRASurgery(PEFTBase):
+    """Operates on the attention layers for performing low-rank adaptation.
+    """
+    def __call__(
```

Comment:
Use `__init__` instead, set all the relevant weights as members (`w_a_linear` etc.).  If needed implement a `get_qkv` function that you can call in `PEFT_Sam` to do the block update.